exercises/tom_hampson/ex01_basics/src/main.cpp

```diff
@@ -0,0 +1,189 @@
+    unsigned int paddingSize = 1 + longestWordLength + maxDigits - static_cast<int>(wordCount.first.length());
+    outputFile << wordCount.first << std::setw(paddingSize) << wordCount.second << std::endl;
+  }
+
+  outputFile.close();
+
+  return true;
+}
+
+/** Read text file and count instances of words with >= 4 letters, then save the information to a file.
+ * @param argc
+ * @param argv - user must supply an input file name, and optionally an output file name.
+ * @return
+*/
+int main(int argc, char* argv[])
+{
+  //Check that there are either 2 or 3 arguments. User must specify an input file name,
+  //and optionally they can specify and output file name.
+  if (argc < 2 || argc > 3)
+  {
+    std::cout << "Please provide arguments for input file (and optionally output file)." << std::endl;
+    return 1;
+  }
+
+  //Input file is the first argument.
+  std::string inputFileName = argv[1];
+
+  //If there was an output file name specified, use that. Otherwise default to output.txt.
+  std::string outputFileName = "output.txt";
+  if (argc == 3)
+  {
+    outputFileName = argv[2];
+  }
+
+  std::ifstream inputFile(inputFileName);
+
+  if (!inputFile)
+  {
+    std::cout << "Unable to open file " << inputFileName << std::endl;
+    return 1;
+  }
+
+  //Line of text from input source.
+  std::string line;
+
+  std::map<std::string, int> wordCounts;
+
+  //Read file one line at a time, or until we reach a hyphen.
+  while (std::getline(inputFile, line, '-'))
+  {
+    std::istringstream wordsInLine(line);
+
+    std::string word;
+
+    //Loop over words in the line, using whitespace as delimiter.
+    while (wordsInLine >> word)
+    {
+      //Only consider words > 3 characters long.
+      if (word.length() < 4)
+      {
+        continue;
+      }
+      //Remove all punctuation characters:
+      //.,?'"!():
+      //Need to escape several of them because they are special pattern
+      //characters in regex expressions.
+      std::regex specialCharacters("\\.|,|\\?|'|\"|!|\\(|\\)|:");
+      std::string newWord = std::regex_replace(word, specialCharacters, "");
+
+      //Check again for the length of the word because it will be shorter if punctuation was removed.
+      if (newWord.length() < 4)
+      {
+        continue;
+      }
+
+      //Convert word to lower case.
+      std::transform(newWord.begin(), newWord.end(), newWord.begin(), [](char c) { return static_cast<char>(std::tolower(c)); });
+
+      //Increment the count of the word.
+      wordCounts[newWord]++;
+    }
+  }
+
+  //If there are no words, don't do anything else.
+  if (wordCounts.size() == 0)
+  {
+    std::cout << "No valid words found in file " << inputFileName << std::endl;
+    return 1;
+  }
+
+  //Container to store the word counts sorted by count.
+  std::vector<std::pair<std::string, int>> sortedWordCounts;
+
+  std::copy(wordCounts.begin(), wordCounts.end(), std::back_inserter(sortedWordCounts));
```

Comment:
Could this be: 
```suggestion
  std::move(wordCounts.begin(), wordCounts.end(), std::back_inserter(sortedWordCounts));
```
Doesn't look like we use the wordCounts after these lines, so the move will save us copying each pair.

Reply:
Done